More generic mech construction messages & balloon alerts

dragomagol opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
The mech construction code was really bad. So much copy paste code. I came in to make them into balloon alerts and then ended up by torturing coders by showing them the hot garbage that was within. The messages will be mostly the same, but now the mech construction steps all use the same code to display user feedback.

The HONK mech's is still not great, but it doesn't really fit into the same mold as the other mechs' construction.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
More balloon alerts for construction tasks, less copy/paste code.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
qol: mech construction now has balloon alerts!
code: mecha_construction_paths.dm is now has significantly less copy/paste code
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Commit:
More generic mech construction messages

MrMelbert commented on code/modules/vehicles/mecha/mecha_construction_paths.dm:
```diff
@@ -790,33 +585,31 @@
 
 	//TODO: better messages.
 	switch(index)
 		if(2)
-			user.visible_message(span_notice("[user] installs [I] into [parent]."), span_notice("You install [I] into [parent]."))
+			user.balloon_alert_to_viewers("installed [I]")
```
If theese are all the same we can just cut down on duplication anyways

```
switch(index)
   if(2, 4, 6, 8, 10, 12)
      user.balloon_alert_to_viewers("installed [I]")
   if(14, 16)
      user.balloon_alert_to_viewers("added [I]")

MrMelbert commented on code/modules/vehicles/mecha/mecha_construction_paths.dm:
```diff
@@ -1072,138 +754,13 @@
-					span_notice("You slowly place the bluespace anomaly core into its socket and close its chamber."))
+	// We've already advanced the message step in ..(), so we have to offset by one to make sure we're on the right message
+	if((message_step - 1) == INSERT_ANOMALY_CORE_STEP)
+		if(diff==FORWARD)
```
```suggestion
		if(diff == FORWARD)
```

MrMelbert commented on code/modules/vehicles/mecha/mecha_construction_paths.dm:
```diff
@@ -1072,138 +754,13 @@
+	// We've already advanced the message step in ..(), so we have to offset by one to make sure we're on the right message
+	if((message_step - 1) == INSERT_ANOMALY_CORE_STEP)
+		if(diff==FORWARD)
+			user.balloon_alert_to_viewers("inserted and secured bluespace anomaly core")
```
Might be a little lengthy for a `balloon_alert_to_viewers`

dragomagol replied:
Yeah... it's kind of two steps in one. I think chopping off the "and secured" would be easiest.

MrMelbert commented on code/modules/vehicles/mecha/mecha_construction_paths.dm:
```diff
@@ -263,6 +300,89 @@
+
+	if(!has_weapons_module)
+		if(curr_step == ADD_WEAPONS_CONTROLS_STEP || curr_step == SECURE_WEAPONS_CONTROLS_STEP)
+			message_step += diff*2
+			curr_step += diff*2
```
```suggestion
			message_step += diff * 2
			curr_step += diff * 2
```

MrMelbert commented on code/modules/vehicles/mecha/mecha_construction_paths.dm:
```diff
@@ -263,6 +300,89 @@
+
+	if(!has_bluespace_crystal)
+		if(curr_step == INSTALL_BLUESPACE_STEP || curr_step == CONNECT_BLUESPACE_STEP || curr_step == ENGAGE_BLUESPACE_STEP)
+			message_step += diff*3
+			curr_step += diff*3
```
```suggestion
			message_step += diff * 3
			curr_step += diff * 3
```

MrMelbert reviewed (commented):
Cool stuff

dragomagol committed:
Update code/modules/vehicles/mecha/mecha_construction_paths.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

dragomagol committed:
Update code/modules/vehicles/mecha/mecha_construction_paths.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

dragomagol committed:
Update code/modules/vehicles/mecha/mecha_construction_paths.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

Commit:
Suggestions

Commit:
forgot one undef

zxaber commented:
Did the steps change at all from the players' view?

dragomagol commented:
> Did the steps change at all from the players' view?

The biggest change was making the external plating step be generic "installed external armor" instead of "installed Gygax armor plates"

LemonInTheDark commented:
Love you brother

LemonInTheDark commented on code/modules/vehicles/mecha/mecha_construction_paths.dm:
```diff
@@ -263,6 +300,89 @@
+	var/curr_step = forward ? message_step : (message_step - 1)
+
+	// Skip over steps we're not doing!
+	if(!has_treads && curr_step == ADD_TREADS_STEP)
+		message_step += diff
+		curr_step += diff
+
+	if(!has_weapons_module)
+		if(curr_step == ADD_WEAPONS_CONTROLS_STEP || curr_step == SECURE_WEAPONS_CONTROLS_STEP)
+			message_step += diff * 2
+			curr_step += diff * 2
+
+	if(!has_bluespace_crystal)
+		if(curr_step == INSTALL_BLUESPACE_STEP || curr_step == CONNECT_BLUESPACE_STEP || curr_step == ENGAGE_BLUESPACE_STEP)
+			message_step += diff * 3
+			curr_step += diff * 3
```
add a comment here about the order being important, since it is important

dragomagol replied:
I'm not sure exactly which part you mean. The order of the if blocks shouldn't matter here, since it's dependent on the step we're on.

LemonInTheDark replied:
Lets say we're on ADD_TREADS_STEP, and we don't have treads.
Pretending for a moment there was another optional buildstep directly after treads, we enter that step.
If we checked for that step before we checked for treads, we would be in an illegal position

dragomagol replied:
What would happen if we skipped backwards into a step we didn't want? Wouldn't that have the same issue but the steps are out of order?

LemonInTheDark replied:
yes actually, it seems to be kinda an issue with how your logic is setup.
maybe have a get_next_step() proc  with a while loop or something?

dragomagol replied:
The way that construction code works for messages pains me greatly, but that's probably a better idea.

dragomagol committed:
Protect against skipping backwards and forwards steps

This is a webedit only because I'm currently getting fucked by python dependencies and can't use git bash to commit - this has been tested

dragomagol committed:
please work webedit gods

Commit:
Fix lint

LemonInTheDark commented on code/modules/vehicles/mecha/mecha_construction_paths.dm:
```diff
@@ -252,7 +289,100 @@
+	var/on_valid_step = FALSE
+
+	while (!on_valid_step)
+		var/next_step = forward ? message_step : (message_step - 1)
```
why `(message_step - 1)` here exactly? I don't really understand.

dragomagol replied:
Construction code is funky when it comes to messages. I had to write it out a couple times before I got it.

Let's ignore the treads step for this example.

We start on step 1. We haven't changed any numbers yet, and we're moving forward in construction, so we tell the user that they've connected the hydraulics systems, and advance to step 2.

Now if we want to disconnect the hydraulics systems, we can't print and then decrement the number; that would tell the user that they've deactivated the hydraulics instead of disconnecting them. That's why we have to offset the backwards step by one, in order to condense the code displaying messages.

Otherwise you'd get something like the old code:

```
if(2)
	if(diff==FORWARD)
		user.visible_message(span_notice("[user] activates [parent] hydraulic systems."), span_notice("You activate [parent] hydraulic systems."))
	else
		user.visible_message(span_notice("[user] disconnects [parent] hydraulic systems."), span_notice("You disconnect [parent] hydraulic systems."))
```

dragomagol replied:
I'll make a better comment for it in code

Commit:
Better comments
